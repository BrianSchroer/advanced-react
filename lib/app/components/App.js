import React from 'react';
import PropTypes from 'prop-types';
import glamorous from 'glamorous';
import Perf from 'react-addons-perf';
import pickBy from 'lodash.pickby';
import { ArticleList } from '../../modules/articles';
import SearchBar from './SearchBar'; // eslint-disable-line import/no-named-as-default
import Timestamp from './Timestamp'; // eslint-disable-line import/no-named-as-default

if (typeof window !== 'undefined') {
  window.Perf = Perf;
}

const Main = glamorous.main({
  fontFamily: '"Segoe UI", Tahoma, Geneva, Verdana, sans-serif',
  maxWidth: '1200px',
  margin: '0 auto'
});

export default class App extends React.PureComponent {
  static childContextTypes = {
    store: PropTypes.object
  };

  static propTypes = {
    store: PropTypes.object.isRequired
  };

  //eslint-disable-next-line react/sort-comp
  appState = () => {
    const { articles, searchTerm } = this.props.store.getState();
    return { articles, searchTerm };
  };

  state = this.appState();

  getChildContext() {
    return { store: this.props.store };
  }

  diagnosePerformance() {
    setTimeout(() => {
      Perf.start();
    }, 0);

    setTimeout(() => {
      Perf.stop();
      Perf.printWasted();
    }, 5000);
  }

  componentDidMount() {
    const { store } = this.props;
    this.subscriptionId = store.subscribe(this.onStoreChange);
    store.startClock();

    // this.diagnosePerformance();
  }

  componentWillUnmount() {
    this.props.store.unsubscribe(this.subscriptionId);
  }

  onStoreChange = () => {
    this.setState(this.appState());
  };

  render() {
    const { searchTerm } = this.state;
    let { articles } = this.state;

    if (searchTerm) {
      const caseInsensitiveRegExp = new RegExp(searchTerm, 'i');

      articles = pickBy(
        articles,
        a =>
          a.title.match(caseInsensitiveRegExp) ||
          a.body.match(caseInsensitiveRegExp)
      );
    }

    return (
      <Main>
        <Timestamp />
        <SearchBar />
        <ArticleList articles={articles} />
      </Main>
    );
  }
}
