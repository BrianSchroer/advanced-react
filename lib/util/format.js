export default class {
  static date(dateString) {
    return new Date(dateString).toDateString();
  }

  static dateTime(dateTimeString) {
    const dt = new Date(dateTimeString);
    return `${dt.toDateString()} ${dt.toLocaleTimeString([], {
      hour: '2-digit',
      minute: '2-digit'
    })}`;
  }
}
