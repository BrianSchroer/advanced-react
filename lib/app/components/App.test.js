import React from 'react';
import { snapshotHelper } from '../../util/testHelpers';
import App from './App';

const testProps = {
  store: {
    getState: () => ({
      articles: {},
      authors: {},
      searchTerm: '',
      timestamp: new Date('Mon Jul 11 2016 00:00:00 GMT+0000 (UTC)')
    }),
    subscribe: () => {},
    unsubscribe: () => {},
    startClock: () => {},
    setSearchTerm: () => {}
  }
};

describe('App', () => {
  it('renders correctly', () => {
    snapshotHelper.assertMatch(<App {...testProps} />);
  });
});
