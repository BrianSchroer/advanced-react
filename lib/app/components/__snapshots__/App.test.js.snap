// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<main
  className="css-boyygm"
>
  <div>
    Sun Jul 10 2016 19:00
  </div>
  <input
    className="css-28xugj"
    onChange={[Function]}
    placeholder="Enter search term"
    type="search"
    value=""
  />
  <div />
</main>
`;
